# To inform the algorithm to exit automatically when no data is received
autoexit:           1

# Declare the lidar topics to merge
lidar_topic:        [
                      "/livox/lidar_ouster"
                      # ,
                      # "/os_cloud_node/points"                      
                    ]

merged_lidar_topic: "/merged_pc"

# Lidar extrinsics: Transform of coordinates in lidar's frame to body
# NOTE TO USER: The (4,4) entry of the extrinsics is used to declare
# whether stamp is at beginning (1.000) or end (0.000) of scan
lidar_extr:         [
                      # Livox
                      0.99985813160508060000, -0.0002258196228773494, -0.0168423771687589430, -0.010514839241742317,
                     -0.00017111407692153792, -0.9999947058517530000,  0.0032494597148797887, -0.008989784841758377,
                     -0.01684302179448474500, -0.0032461167514232880, -0.9998528768488223000,  0.037356468638334630,
                      0.00000000000000000000,  0.000000000000000000,   0.0000000000000000000,  1.000000000000000000
                    #   ,
                    #   # Ouster
                    #   0.999934655205122900,    0.003477624535771754,  -0.010889970036688295,  -0.060649229060416594,
                    #   0.003587143302461965,   -0.999943027982117100,   0.010053516443599904,  -0.012837544242408117,
                    #  -0.010854387257665576,   -0.010091923381711220,  -0.999890161647627000,  -0.020492606896077407,
                    #   0.000000000000000000,    0.000000000000000000,   0.000000000000000000,   0.000000000000000000
                    ]

# Minimum lidar range to admit
min_range:          1.0

# IMU topic
imu_topic:          "/vn100/imu"

# Gravity constant
GRAV:               9.82

# IMU noises
GYR_N:              20
GYR_W:              10
ACC_N:              1
ACC_W:              10

# Position prior weight
POSE_N:             100

# Noise of velocity propagation
VEL_N:              10

# Weightage of a basic lidar factor
lidar_weight:       10

# Bias bounds
BG_BOUND:           [ 0.1,  0.1,  0.2]
BA_BOUND:           [ 0.1,  0.1,  0.2]

# Number of packets to process at time
WINDOW_SIZE:        3
# Number of substates to give to each packet
N_SUB_SEG:          4
# Spline order
SPLINE_N:           4
# Knot length
deltaT:             0.025
# Period at the beginning of the spline to fix
start_fix_span:    -1.0
# Period at the end of the spline to fix
final_fix_span:    -1.0
# Steps to reassociate
reassociate_steps:  1
# Deskew choice, 0 for IMU prop, 1 for SPLINE, 1 for before and 1 for after
deskew_method:      [ 0, 0 ]
# Rate of doing reassociation
reassoc_rate:       0

# Downsampling grid size, chose -1 for no downsampling
leaf_size:          0.2           # Downsample scale of input pointcloud, also the scale of ufomap leaf
assoc_spacing:     -0.2
surfel_map_depth:   16
surfel_min_point:   6
surfel_min_depth:   1
surfel_query_depth: 6             # Equal to Dmax + 1. Dmax is defined in the paper
surfel_intsect_rad: 0.1           # r in the the paper
surfel_min_plnrty:  0.7
use_ufm:            1

# Distance to surfel and min score thresholds
dis_to_surfel_max:  0.1
score_min:          0.0

# Downsample coefficient of lidar scans, ds_rate = n means keep every n-th point.
ds_rate:            [
                      1
                    ]

# Number of pointclouds to merge into one
sweep_len:          1

# Threshold to add new keyframe
kf_min_dis:         0.5
kf_min_angle:       10
refine_kf:          0

# Optimization parameters
lidar_loss_thres:  -10.0
imu_loss_thres:    -1.0

# Solver:
# Linear solver: "dnc" "snc" "cgnr" "dschur" "sschur" "ischur"
linSolver:          "snc"
# Trust region strategy: "lm", "dogleg"
trustRegType:       "lm"
# Linear Algebra library used: "eigen", "lapack", "cuda"
linAlgbLib:         "eigen" 
max_solve_time:     0.3
max_iterations:     1
ensure_real_time:   0
find_factor_cost:   0
fit_spline:         0

# Sensors fused
fuse_lidar:         1
fuse_imu:           1
fuse_poseprop:      0
fuse_velprop:       0

regularize_imu:     0
imu_init_time:      0.5
max_outer_iters:    6
max_lidar_factor:   4000    # Maximum number of lidar factors
dj_thres:           0.05    # Threshold to break the outer iterations early

# Loop closure parameters
loop_en:            1
loop_kf_nbr:        20    # Number of neighbour to check for loop closure
loop_kfid_mindiff:  20    # Only check for loop when keyframes have this much difference

icpMaxIter:         100   # Maximum iterations for ICP
icpFitnessThres:    0.6   # Fitness threshold for ICP check
histDis:            15    # Maximum correspondence distance for icp

# Bundle adjustment params
rib_edge:           100
odom_q_noise:       0.1
odom_p_noise:       0.1
loop_weight:        0.05

# Number of optimizations before quitting
publish_map:        0

# Terminal outputs
show_report:        1
show_preop_times:   1
show_lioop_times:   1

# Log directory
log_dir:            "/home/tmn/slict_log/atv"

# Configurations for relocalization
relocalization:

  # pre process
  ds_size:                    0.25
  maximum_corner_num:         100

  # key points
  plane_detection_thre:       0.01
  plane_merge_normal_thre:    0.2
  voxel_size:                 2.0
  voxel_init_num:             10
  proj_image_resolution:      0.5
  proj_dis_min:               0
  proj_dis_max:               5
  corner_thre:                10

  # std descriptor
  descriptor_near_num:        10
  descriptor_min_len:         2
  descriptor_max_len:         50
  non_max_suppression_radius: 2
  std_side_resolution:        0.2

  # candidate search
  skip_near_num:              100
  candidate_num:              50
  sub_frame_num:              1
  vertex_diff_threshold:      0.5
  rough_dis_threshold:        0.01
  normal_threshold:           0.2
  dis_threshold:              0.5
  icp_threshold:              0.4

  # For guestimation filtering
  refine_reloc_tf:            0
  reloc_thres:                3
  reloc_timespan:             5.0
  pose_percentile_thres:      0.7
  priormap_viz_res:           0.2
  marginalize_new_points:     0
